### ISSUE TITLE
Intersecting generic type with itself leaks protected methods

### ISSUE THREAD

**Issue by flapenguin** [Microsoft/TypeScript#27716]
**TypeScript Version:**  3.2.0-dev.20181011

**Search Terms:** generic protected intersection leak

**Code**

```ts
class GenericSecret<T extends {}> {
    private privateMethod() { }
    protected protectedField: any;
    protected protectedMethod() { }
    protected protectedGenericMethod<TKey extends keyof T>(key: TKey) { }
}

type Broken = GenericSecret<{ a: void; }> & GenericSecret<{ a: void; }>;
({} as Broken).privateMethod();             // expected: error  actual: error
({} as Broken).protectedField;              // expected: error  actual: error
({} as Broken).protectedMethod();           // expected: error  actual: no error <<<<<< 
({} as Broken).protectedGenericMethod('a'); // expected: error  actual: no error <<<<<<
```

**Expected behavior**: accessing protected methods generates error

**Actual behavior**: no error

**Playground Link:**  [link](http://www.typescriptlang.org/play/index.html#src=class%20GenericSecret<T%20extends%20%7B%7D>%20%7B%0D%0A%20%20%20%20private%20privateMethod()%20%7B%20%7D%0D%0A%20%20%20%20protected%20protectedField%3A%20any%3B%0D%0A%20%20%20%20protected%20protectedMethod()%20%7B%20%7D%0D%0A%20%20%20%20protected%20protectedGenericMethod<TKey%20extends%20keyof%20T>(key%3A%20TKey)%20%7B%20%7D%0D%0A%7D%0D%0A%0D%0Atype%20Broken%20%3D%20GenericSecret<%7B%20a%3A%20void%3B%20%7D>%20%26%20GenericSecret<%7B%20a%3A%20void%3B%20%7D>%3B%0D%0A(%7B%7D%20as%20Broken).privateMethod()%3B%20%20%20%20%20%20%20%20%20%20%20%20%20%2F%2F%20expected%3A%20error%20%20actual%3A%20error%0D%0A(%7B%7D%20as%20Broken).protectedField%3B%20%20%20%20%20%20%20%20%20%20%20%20%20%20%2F%2F%20expected%3A%20error%20%20actual%3A%20error%0D%0A(%7B%7D%20as%20Broken).protectedMethod()%3B%20%20%20%20%20%20%20%20%20%20%20%2F%2F%20expected%3A%20error%20%20actual%3A%20no%20error%20<<<<<<%20%0D%0A(%7B%7D%20as%20Broken).protectedGenericMethod('a')%3B%20%2F%2F%20expected%3A%20error%20%20actual%3A%20no%20error%20<<<<<<%0D%0A)

**Related Issues:**

**Comment by flapenguin** [comment]
I've digged a little. Symbol generated for `protectedMethod` in `createUnionOrIntersectionProperty` does not have `ModifierFlags.NonPublicAccessibilityModifier`. So `checkPropertyAccessExpressionOrQualifiedName` is ok with it.

**Comment by flapenguin** [comment]
Found workaround:

```ts
class HackityHack {
    protected protectedMethod(whatever: never) { }
}

class GenericSecret<T extends {}> {
    protected protectedMethod<TT extends keyof T>(key: TT) { }
}

type NotBroken = HackityHack & GenericSecret<{ a: void; }> & GenericSecret<{ a: void; }>;
({} as NotBroken).protectedMethod('a') // yay, error
```

**Comment by DanielRosenwasser** [comment]
@flapenguin would you be willing to send a PR since you dug through the source and have a minimal repro? Great way to help the compiler team and potentially get a Hacktoberfest t-shirt 😄

**Comment by flapenguin** [comment]
@DanielRosenwasser yes, I will create the PR.

**Comment by Kingwl** [comment]
> Great way to help the compiler team and potentially get a Hacktoberfest t-shirt 😄  

I also want a T-shirt🙋🏻‍♂️.

**Comment by Kingwl** [comment]
I also want a T-shirt🙋🏻‍♂️.

**Comment by Kingwl** [comment]
I also want a T-shirt🙋🏻‍♂️.
